cores/esp8266/Esp.cpp

```diff
@@ -138,6 +138,186 @@ uint64_t EspClass::deepSleepMax()
+    auto ret_do_sleep = wifi_fpm_do_sleep(duration_us);
+    if (ret_do_sleep != 0)
+    {
+#ifdef DEBUG_SERIAL
+        DEBUG_SERIAL.printf("core: error %d with wifi_fpm_do_sleep: (-1=sleep status error, -2=force sleep not enabled)\n", ret_do_sleep);
+#endif
+        return false;
+    }
+    // SDK turns on forced modem sleep in idle task
+#ifdef HAVE_ESP_SUSPEND
+    esp_delay(10);
+#else
+    delay(10);
+#endif
+    return true;
+}
+
+void EspClass::forcedModemSleepOff()
+{
+    const sleep_type_t sleepType = wifi_fpm_get_sleep_type();
+    if (sleepType != NONE_SLEEP_T) {
+        if (sleepType == MODEM_SLEEP_T) wifi_fpm_do_wakeup();
+        wifi_fpm_close();
+    }
+    wifi_fpm_set_sleep_type(saved_sleep_type);
+    saved_sleep_type = NONE_SLEEP_T;
+}
+
+#ifdef DEBUG_SERIAL
+namespace {
+    void walk_timer_list() {
+        os_timer_t* timer_root;
+        {
+            esp8266::InterruptLock lock;
+            auto src = timer_list;
+            auto dest = src ? new os_timer_t : nullptr;
+            timer_root = dest;
+            while (dest) {
+                *dest = *src;
+                src = src->timer_next;
+                dest->timer_next = src ? new os_timer_t(*timer_list) : nullptr;
+                dest = dest->timer_next;
+            }
+        }
+        DEBUG_SERIAL.printf("=============\n");
+        for (os_timer_t* timer_node = timer_root; nullptr != timer_node; timer_node = timer_node->timer_next) {
+            DEBUG_SERIAL.printf("timer_address = %p\n", timer_node);
+            DEBUG_SERIAL.printf("timer_expire  = %u\n", timer_node->timer_expire);
+            DEBUG_SERIAL.printf("timer_period  = %u\n", timer_node->timer_period);
+            DEBUG_SERIAL.printf("timer_func    = %p\n", timer_node->timer_func);
+            DEBUG_SERIAL.printf("timer_next    = %p\n", timer_node->timer_next);
+            if (timer_node->timer_next) DEBUG_SERIAL.printf("=============\n");
+        }
+        DEBUG_SERIAL.printf("=============\n");
+        DEBUG_SERIAL.flush();
+        while (timer_root) {
+            auto next = timer_root->timer_next;
+            delete timer_root;
+            timer_root = next;
+        }
+    }
+}
+#endif
+
+bool EspClass::forcedLightSleepBegin(uint32_t duration_us, fpm_wakeup_cb wakeupCb)
+{
+    // Setting duration to 0xFFFFFFF, it disconnects the RTC timer
+    if (!duration_us || duration_us > 0xFFFFFFF) {
+        duration_us = 0xFFFFFFF;
+    }
+    wifi_fpm_close();
+    saved_sleep_type = wifi_fpm_get_sleep_type();
+    wifi_set_opmode(NULL_MODE);
+    wifi_fpm_set_sleep_type(LIGHT_SLEEP_T);
+    wifi_fpm_open();
+    saved_wakeupCb = wakeupCb;
+    wifi_fpm_set_wakeup_cb([]() {
+        if (saved_wakeupCb) {
+            saved_wakeupCb();
+            saved_wakeupCb = nullptr;
+        }
+        esp_schedule();
+        });
+#ifdef DEBUG_SERIAL
+    walk_timer_list();
+#endif
+    {
+        esp8266::InterruptLock lock;
+        saved_timer_list = timer_list;
```

Comment:
Repeating the q from above - what is the difference between GPIO-activated sleep and timed one here? Is CPU really sleeping or we just simulate that it does?

Reply:
OMG, you are asking questions that I cannot remember the answer to, except that I am 100% certain that we literally did hours of repetitive tests to show that power consumption dropped to the expected levels, and it didn't or the MCU became unreliable or the tables got corrupted if it wasn't done in exactly this - documented by Espressif - way. We really should not change a single line.

Reply:
Right, but now we have a half-understood feature that maybe works?
Reading our issues simply repeats the above - we copy nodemcu and test how it did. Would have to read that yet again if you don't have an explanation :/
(https://github.com/nodemcu/nodemcu-firmware/pull/1231#issuecomment-314497558 and below)

Reply:
I'm not sure we are not talking past each other. The sleep isn't actived by the GPIO, but cancelled. Same for the timer. I'm sure that's what you meant. As far as it got measure two years ago, yes we were sure it did as advertised - sleeping to save power, resuming either by GPIO trigger or timeout.

Reply:
@mcspr Is this now stalled again due to to unrelated workload, or have my changes and explanations been unconvincing so far? I would be totally relieved if this finally got merged, I am sure it will help people a lot, at least it might allow more specific reports about power-saving mode issues?